Gh-275: Schema Structure documentation (#377)
gchq/gaffer-doc: docs/getting-started/advanced-guide/schema.md

The sections below walkthrough the features of Schemas in detail and explain how to accurately model your data in Gaffer.

## Elements schema
The Elements schema is designed to be a high level document describing what information your Graph contains, i.e. the different kinds of edges and entities and the list of properties associated with each.
Essentially this part of the schema should just be a list of all the entities and edges in the graph. 
Edges describe the relationship between a source vertex and a destination vertex. 
Entities describe a vertex. We use the term "element" to mean either and edge or and entity.

When defining an element we must provide a "group". This is a unique string identifier for each element.
Groups must be completely unique and cannot be shared between edges and entities.

### Edges
Edges must have the following:

- `source` - The type of object that will be used as the source vertex in your graph. Needs to be a type defined in the [Types schema](#types).

> - `destination` - Similar to source, it can either be the same or a different type.

Reviewer: Potentially should mention that this is the destination vertex, just to ensure clarity.

Author: I've reworded this.

- `directed` - Tells Gaffer if the edge is directed or undirected. Needs to be a type which means true or false, see [Types](#true--false) for more info.

> When an Edge is undirected in Gaffer (`directed` is `false`), it is treated as if the relationship was bidirectional and the vertices of the edge do not have an authoritative source and destination thus, the undirected edges `A -- B` and `B -- A` are equal, and therefore will be aggregated together.

Reviewer: ```suggestion
When an Edge is undirected in Gaffer (`directed` is `false`), it is treated as if the relationship was bidirectional and the vertices of the edge do not have an authoritative source and destination. Thus, the undirected edges `A -- B` and `B -- A` are equal, and therefore will be aggregated together.
```

Gaffer will present the undirected edges Vertices in natural ordering, so a client will also see the above edge presented as `A, B`.
It will be aggregated with any other edge that has the same source and destination and is also undirected.
For example an undirected Edge that is added, `A -> B`, will be aggregated with another Edge of `B -> A` if they are undirected.

### Entities
Entities must have a `vertex` field, which is similar to the `source` and `destination` fields on an edge.
For example, modelling `London -> Paris` requires an entity definition to represent the city and an edge definition to represent the relationship.

> In that example the entity `vertex` field and the edge `source` and `destination` fields would all be of the same type.

Reviewer: ```suggestion
In this example the entity `vertex` field and the edge `source` and `destination` fields would all be of the same type.
```

Author: I'm not sure about this. I added lines 38 & 39 to provide an example, but this example is separate from the main example schema discussed on the page. Which is why I wrote 'that' rather than 'this'. Maybe it would be better to rephrase in general?

Reviewer: I had another read and I am still not sure - I think maybe rephrase? I feel the 'that' makes it less clear that you mean the example in the previous sentence.

Author: I've reworded this to refer to the previous example to avoid any confusion.

### Optional Element fields
Edges and Entities can optionally have the following fields:

- `description` - A simple string which should provide some context and explain what the element is.
- `parents` - An array of parent group names. These must relate to the same sort of element as the child, for example an edge cannot have an entity as a parent. Elements can inherit any information from multiple parent elements. Fields will be merged/overridden, so the hierarchy of parents is important. Any fields that are defined in the child element will also merge or override information taken from the parents.
- `properties` - Properties are defined by a map of key-value pairs of property names to property types. Property types are described in the Types schema.
- `groupBy` - By default Gaffer uses the element group and its vertices to group similar elements together when aggregating and summarising elements. This groupBy field allows you to specify extra properties (in addition to the element group and vertices) to use for controlling when similar elements should be grouped together and summarised.

> - `visibilityProperty` - When using visibility properties in your graph, ensure the sensitive elements have a visibility property and then set this visibilityProperty field to that property name. This ensures Gaffer knows to restrict access to sensitive elements.

Reviewer: I would potentially swap the sentence order, you usually would want the quick overview of what the property is before the explanation of how to implement it.

Author: I've reworded/reordered this. Should now be clearer what each property is first, before going into more detailed.

> - `timestampProperty` - When using the timestamp property in your graph, then set this timestampProperty field to that property name, so Gaffer Stores know to treat that property specially. This property allows Store implementations like Accumulo to optimise the way the timestamp property is persisted. For these stores using it can have a very slight performance improvement due to the lazy loading of properties. Setting the timestampProperty is optional and does not affect the queries available to users. For more information [see the timestamp section of the Accumulo Store Reference](../../reference/stores-guide/accumulo.md#timestamp).

Reviewer: I feel like this is missing a really simple explanation of what the timestamp property is.

      "class" : "java.util.Date",
      "aggregateFunction" : {
        "class" : "uk.gov.gchq.koryphe.impl.binaryoperator.Max"
      },
      "validateFunctions" : [ {
        "class" : "uk.gov.gchq.koryphe.impl.predicate.Exists"
      } ]
    },
    "set" : {
      "class" : "java.util.TreeSet",
      "aggregateFunction" : {
        "class" : "uk.gov.gchq.koryphe.impl.binaryoperator.CollectionConcat"
      }
    },
    "hllSketch" : {
      "class" : "org.apache.datasketches.hll.HllSketch",
      "aggregateFunction" : {
        "class" : "uk.gov.gchq.gaffer.sketches.datasketches.cardinality.binaryoperator.HllSketchAggregator"
      },
      "serialiser" : {
        "class" : "uk.gov.gchq.gaffer.sketches.datasketches.cardinality.serialisation.HllSketchSerialiser"
      }
    }
  }
}
```

## Java API

Schemas can be loaded from a JSON file directly using the [`fromJSON()` method of the `Schema` class](https://gchq.github.io/Gaffer/uk/gov/gchq/gaffer/store/schema/Schema.html#fromJson(java.io.InputStream...)). This accepts `byte[]`, `InputStream` or `Path` types, for example:

```java
Schema mySchema = Schema.fromJson(Paths.get("path/to/schema.json"));
```

> While it's easiest to create Schemas objects using the JSON approach, you can create them directly from their component classes in Java.

Reviewer: ```suggestion
While it's easiest to create Schema objects using the JSON approach, you can create them directly from their component classes in Java.
```

# Schemas
This page delves into more detail on Schemas. As seen in the other guides, there are two main components to schemas: Elements and Types.

Schemas are often separated into two JSON files (`elements.json` & `types.json`), but it's worth remembering that your schema can be broken up further if required.

For example if you have a Graph that is made up of 5 different data sources, you may find it easier to develop and maintain by splitting your schema into multiple parts, for example 5 Element Schemas and 1 Types schema.

> Alternatively, a single overall schema file could be used for Elements and Types. There's also no requirement to have a single Types schemas.

Reviewer: ```suggestion
Alternatively, a single overall schema file could be used for Elements and Types, as there is no requirement to have a single Types schemas.
```

When you construct your Gaffer graph you must provide all the Schema parts. If provided separately, these will then be merged together to form an internal schema used for the Graph.

When using Java directly, you don't need to use JSON files to store you schemas - although this is still recommended. See the [Java API](#java-api) section for more about interacting with Schemas in Java.

The sections below walkthrough the features of Schemas in detail and explain how to accurately model your data in Gaffer.

## Elements schema
The Elements schema is designed to be a high level document describing what information your Graph contains, i.e. the different kinds of edges and entities and the list of properties associated with each.
Essentially this part of the schema should just be a list of all the entities and edges in the graph. 
Edges describe the relationship between a source vertex and a destination vertex. 

> Entities describe a vertex. We use the term "element" to mean either and edge or and entity.

Reviewer: ```suggestion
Entities describe a vertex. Edges describe the relationship between a source vertex and a destination vertex.
We use the term "element" to mean either an edge or an entity.
```

I would swap these just so it flows a little better and also just fixed the grammar.

When defining an element we must provide a "group". This is a unique string identifier for each element.
Groups must be completely unique and cannot be shared between edges and entities.

### Edges
Edges must have the following:

- `source` - Type of object to use as the source vertex in your graph. Needs to be a type defined in the [Types schema](#types).
- `destination` - Type of object to use as the destination vertex in your graph. Can either be the same type as `source`, or a different type.
- `directed` - Tells Gaffer if the edge is directed or undirected. Needs to be a type which means true or false, see [Types](#true--false) for more info.

> When an Edge is undirected in Gaffer (`directed` is `false`), it is treated as if the relationship was bidirectional, and the vertices of the edge do not have an authoritative source and destination.

Reviewer: ```suggestion
When an Edge is undirected in Gaffer (`directed` is `false`), it is treated as if the relationship was bidirectional, meaning that the vertices of the edge do not have an authoritative source and destination.
```

Thus, the undirected edges `A -- B` and `B -- A` are equal, and will therefore be aggregated together.

> Gaffer will present the undirected edges Vertices in natural ordering, so a client will also see the above edge presented as `A, B`.

Reviewer: ```suggestion
Gaffer will present the undirected edges vertices in natural ordering, so a client will also see the above edge presented as `A, B`.
```

It will be aggregated with any other edge that has the same source and destination and is also undirected.

> For example an undirected Edge that is added, `A -> B`, will be aggregated with another Edge of `B -> A` if they are undirected.

Reviewer: I feel like this paragraph is a bit repetitive? Instead I would combine a couple of sentences:
Thus, the undirected edges A -- B and B -- A are equal, and will therefore be aggregated with any other edge that has the same source and destination and is also undirected. Gaffer will present the undirected edges vertices in natural ordering, so a client will also see the above edge presented as A, B. This would mean that an undirected Edge that is added, A -- B, will be aggregated with another Edge of B -- A if they are undirected.

Reviewer: I also wasn't sure if the two bottom two examples should be -- instead of -> as they are meant to be undirected

Author: This section could certainly be more concise, I'll reword.

Author: I've make it more like what's suggested and also trimmed it further. That last sentence is probably not needed but it does make things very clear.

These 2 optional fields are for advanced users. They can go in the Elements Schema, however we have split them out into separate Validation and Aggregation Schema files for this page, so the logic doesn't complicate the Elements schema.

- `validateFunctions` - An array of selections and predicates to be applied to an element. This allows you to validate based on multiple properties at once. E.g. check a timestamp property together with a time to live property to check if the element should be aged off. Individual property validation is best done as a validateFunction in the property type definition in Types schema.
- `aggregateFunctions` - An array of selections and binary operators to be applied to an element. This allows you to aggregate based on multiple properties at once. It is important to note that types of properties (groupBy, non-groupBy, visibility) cannot be aggregated in the same aggregate function. The timestamp property is treated as a non-groupBy property. Individual property aggregation is best done as a aggregateFunction in the property type definition in the Types schema.

### Example Elements Schema

> Here is an example Elements schema:

Reviewer 2: Do we still want to use the road traffic example here? It's quite difficult to follow and haven't really explained what a graph with this schema would look like or how it is used

Author: I took it directly from the old docs. I'll have a look at simplifying it or explaining the context.

Reviewer 2: Fair that's understandable, I wasn't sure if it was in scope to come up with another example but maybe just some context around it would help

Author: I've added some context and a link to where the example originates.
